### PR TITLE
[Discover-next] add query enhancements plugin as a core plugin

### PR DESCRIPTION
**Pull request by kavilla**
***NOTE***: Will update this and ~~move some of the stuff that was added if it was a 3rd party plugin~~ and will add justifications based on this rubric: 

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7081#issuecomment-2201575495

Also need to swap the default values around.

### Description

Introduces the addition of a new core plugin called `Query Enhancements`

This plugin seems to be focused on enhancing query capabilities, particularly in areas like PPL (Piped Processing Language), SQL, and natural language query assistance.

#### Plugin Configuration and Setup

The plugin is set up with the necessary configuration files:

- src/plugins/query_enhancements/opensearch_dashboards.json: Defines the plugin's metadata and dependencies.
- src/plugins/query_enhancements/common/config.ts: Contains the configuration schema for the plugin.
- src/plugins/query_enhancements/public/plugin.tsx: The main plugin file that sets up the public-facing part of the plugin.
- src/plugins/query_enhancements/server/plugin.ts: The server-side setup for the plugin.

#### Query Assist Feature

A significant part of the plugin is dedicated to a "Query Assist" feature:

- src/plugins/query_enhancements/public/query_assist/: This directory contains components, hooks, and utilities for the Query Assist feature.
- src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx: A React component for the Query Assist bar.
- src/plugins/query_enhancements/server/routes/query_assist/: Server-side routing for Query Assist functionality.

#### Search Interceptors

The plugin introduces several search interceptors:

- src/plugins/query_enhancements/public/search/ppl_search_interceptor.ts
- src/plugins/query_enhancements/public/search/sql_search_interceptor.ts
- src/plugins/query_enhancements/public/search/sql_async_search_interceptor.ts

These interceptors likely modify or enhance the search functionality for different query languages.

#### Data Source Connection

There's a new feature for data source connections:

- src/plugins/query_enhancements/public/data_source_connection/: This directory contains components and services for managing data source connections.

#### Server-side Search Strategies

The plugin implements server-side search strategies for different query types:

- src/plugins/query_enhancements/server/search/ppl_search_strategy.ts
- src/plugins/query_enhancements/server/search/sql_search_strategy.ts
- src/plugins/query_enhancements/server/search/sql_async_search_strategy.ts

#### Modifications to Existing Files

There are some modifications to existing files outside the new plugin directory:

- src/plugins/data/public/index.ts: Exports new types and interfaces.
- src/plugins/data/public/ui/index.ts: Exports new UI components.
- src/plugins/data/public/ui/query_editor/index.tsx: Modifies the QueryEditor component.

#### Testing

The plugin includes a comprehensive test setup:

- src/plugins/query_enhancements/test/: Contains Jest configuration and setup files for testing.

#### Acknowledgements

Contains work by @sejli @joshuali925 @Swiddis @paulstn @mengweieric 
Thx @ashwin-pc for the description summary

### Issues Related

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6072
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6074
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6075

## Changelog

- feat: add query enhancements plugin as a core plugin

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by xluo-aws** [comment]
@kavilla Do we allow OSD core call plugin APIs?

File: src/plugins/query_enhancements/public/search/sql_async_search_interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  fetchDataFramePolling,
+} from '../../common';
+import { QueryEnhancementsPluginStartDependencies } from '../types';
+import { ConnectionsService } from '../data_source_connection';
+
+export class SQLAsyncSearchInterceptor extends SearchInterceptor {
+  protected queryService!: DataPublicPluginStart['query'];
+  protected aggsService!: DataPublicPluginStart['search']['aggs'];
+  protected indexPatterns!: DataPublicPluginStart['indexPatterns'];
+  protected dataFrame$ = new BehaviorSubject<IDataFrameResponse | undefined>(undefined);
```
Comment (ashwin-pc) [inline src/plugins/query_enhancements/public/search/sql_async_search_interceptor.ts:35]: Where is this used?

File: src/plugins/query_enhancements/public/search/sql_async_search_interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  fetchDataFrame,
+  fetchDataFramePolling,
+} from '../../common';
+import { QueryEnhancementsPluginStartDependencies } from '../types';
+import { ConnectionsService } from '../data_source_connection';
+
+export class SQLAsyncSearchInterceptor extends SearchInterceptor {
+  protected queryService!: DataPublicPluginStart['query'];
+  protected aggsService!: DataPublicPluginStart['search']['aggs'];
+  protected indexPatterns!: DataPublicPluginStart['indexPatterns'];
```
Comment (ashwin-pc) [inline src/plugins/query_enhancements/public/search/sql_async_search_interceptor.ts:34]: Is this used?

File: src/plugins/query_enhancements/server/plugin.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  implements Plugin<QueryEnhancementsPluginSetup, QueryEnhancementsPluginStart> {
+  private readonly logger: Logger;
+  private readonly config$: Observable<SharedGlobalConfig>;
+  constructor(private initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+    this.config$ = initializerContext.config.legacy.globalConfig$;
+  }
+
+  public setup(core: CoreSetup, { data, dataSource }: QueryEnhancementsPluginSetupDependencies) {
+    this.logger.debug('queryEnhancements: Setup');
```
Comment (ashwin-pc) [inline src/plugins/query_enhancements/server/plugin.ts:41]: Do we still need this?

File: src/plugins/query_enhancements/server/plugin.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    const router = core.http.createRouter();
+    // Register server side APIs
+    const client = core.opensearch.legacy.createClient('opensearch_observability', {
```
Comment (ashwin-pc) [inline src/plugins/query_enhancements/server/plugin.ts:44]: How does this work?

File: src/plugins/query_enhancements/server/routes/data_source_connection/routes.ts
Comment (ashwin-pc) [inline src/plugins/query_enhancements/server/routes/data_source_connection/routes.ts:1]: Why are we introducing new routes?

File: src/plugins/query_enhancements/server/search/ppl_search_strategy.ts
Comment (ashwin-pc) [inline src/plugins/query_enhancements/server/search/ppl_search_strategy.ts:1]: We should write tests for the search strategies

File: src/plugins/query_enhancements/server/utils/shim_stats.ts
Comment (ashwin-pc) [inline src/plugins/query_enhancements/server/utils/shim_stats.ts:1]: Whats this for?

**Review by ashwin-pc — COMMENTED** [review]
Overall no major blockers. Had a few comments though

**Comment by kavilla** [comment]
> @kavilla Do we allow OSD core call plugin APIs?

yeah. it's actually as simple data.search

File: src/plugins/query_enhancements/public/data_source_connection/components/connections_bar.tsx
[the remark this answers is not in the document]
Comment (kavilla) [reply src/plugins/query_enhancements/public/data_source_connection/components/connections_bar.tsx:1]: yeah this should be tossed once data set is added

**Comment by ashwin-pc** [comment]
> > @kavilla Do we allow OSD core call plugin APIs?
> 
> yeah. it's actually as simple data.search

Wait, what now? where did this happen?

**Review by ashwin-pc — APPROVED** [review]
My comments will be addressed in a fast follow. Approve=ing this for now so that dependency features can be merged in. The comments here are just for code quality and tests

**Comment by joshuali925** [comment]
diff to fix some code and tests in query assist
```diff
diff --git a/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx b/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx
index 84ac854c980..e87e74ce299 100644
--- a/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx
+++ b/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx
@@ -45,7 +45,7 @@ export const QueryAssistBar: React.FC<QueryAssistInputProps> = (props) => {
     const subscription = props.connectionsService
       .getSelectedConnection$()
       .subscribe((connection) => {
-        dataSourceIdRef.current = connection?.id;
+        dataSourceIdRef.current = connection?.dataSource.id;
       });
     return () => subscription.unsubscribe();
   }, [props.connectionsService]);
diff --git a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx
index ea568959152..41fc36dd71c 100644
--- a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx
+++ b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx
@@ -24,13 +24,10 @@ const httpMock = coreSetupMock.http;
 
 jest.mock('../components', () => ({
   QueryAssistBar: jest.fn(() => <div>QueryAssistBar</div>),
-}));
-
-jest.mock('../components/query_assist_banner', () => ({
   QueryAssistBanner: jest.fn(() => <div>QueryAssistBanner</div>),
 }));
 
-describe.skip('CreateExtension', () => {
+describe('CreateExtension', () => {
   afterEach(() => {
     jest.clearAllMocks();
   });
diff --git a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx
index bd990b57f5a..8d51cb8998c 100644
--- a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx
+++ b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx
@@ -28,7 +28,7 @@ const getAvailableLanguages$ = (
   connectionsService.getSelectedConnection$().pipe(
     distinctUntilChanged(),
     switchMap(async (connection) => {
-      const dataSourceId = connection?.id;
+      const dataSourceId = connection?.dataSource.id;
       const cached = availableLanguagesByDataSource.get(dataSourceId);
       if (cached !== undefined) return cached;
       const languages = await http
@@ -102,6 +102,7 @@ interface QueryAssistWrapperProps {
 
 const QueryAssistWrapper: React.FC<QueryAssistWrapperProps> = (props) => {
   const [visible, setVisible] = useState(false);
+  console.log('❗visible:', visible);
 
   useEffect(() => {
     let mounted = true;
@@ -112,6 +113,7 @@ const QueryAssistWrapper: React.FC<QueryAssistWrapperProps> = (props) => {
       props.http
     ).subscribe((languages) => {
       const available = languages.includes(props.dependencies.language);
+      console.log('setting ',props.invert ? !available : available);
       if (mounted) setVisible(props.invert ? !available : available);
     });
 
diff --git a/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts b/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts
index f22bdb14836..5e31191c9c9 100644
--- a/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts
+++ b/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts
@@ -5,25 +5,29 @@
 
 import { ApiResponse } from '@opensearch-project/opensearch';
 import { ResponseError } from '@opensearch-project/opensearch/lib/errors';
+import { loggerMock } from '@osd/logging/target/mocks';
 import { RequestHandlerContext } from 'src/core/server';
-// // eslint-disable-next-line @osd/eslint/no-restricted-paths
-// import { CoreRouteHandlerContext } from 'src/core/server/core_route_handler_context';
 import { coreMock } from '../../../../../core/server/mocks';
-import { loggerMock } from '@osd/logging/target/mocks';
 import { getAgentIdByConfig, requestAgentByConfig } from './agents';
 
-describe.skip('Agents helper functions', () => {
-  // const coreContext = new CoreRouteHandlerContext(
-  //   coreMock.createInternalStart(),
-  //   httpServerMock.createOpenSearchDashboardsRequest()
-  // );
+describe('Agents helper functions', () => {
   const coreContext = coreMock.createRequestHandlerContext();
   const client = coreContext.opensearch.client.asCurrentUser;
   const mockedTransport = client.transport.request as jest.Mock;
   const context: RequestHandlerContext = {
     core: coreContext,
+    // @ts-ignore
     dataSource: jest.fn(),
-    query_assist: { dataSourceEnabled: false, logger: loggerMock.create() },
+    query_assist: {
+      dataSourceEnabled: false,
+      logger: loggerMock.create(),
+      configPromise: Promise.resolve({
+        enabled: true,
+        queryAssist: {
+          supportedLanguages: [{ language: 'PPL', agentConfig: 'testConfig' }],
+        },
+      }),
+    },
   };
 
   afterEach(() => {
diff --git a/src/plugins/query_enhancements/server/types.ts b/src/plugins/query_enhancements/server/types.ts
index 2ab716b9892..1a5d962e705 100644
--- a/src/plugins/query_enhancements/server/types.ts
+++ b/src/plugins/query_enhancements/server/types.ts
@@ -51,7 +51,10 @@ export interface FacetRequest {
   };
 }
 
-declare module '../../../core/server' {
+// TODO declare it in core changes the type for every reference. we only need
+// this declaration in query_enhancements, but it doesn't seem possible
+// https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4274
+/* declare module '../../../core/server' {
   interface RequestHandlerContext {
     query_assist: {
       logger: Logger;
@@ -64,4 +67,4 @@ declare module '../../../core/server' {
       dataSourceEnabled: boolean;
     };
   }
-}
+} */
```

**Review by abbyhu2000 — APPROVED** [review]

**Comment by kavilla** [comment]
l will breakdown the fast follows as issues and re-link them here.

**Comment by kavilla** [comment]
> diff to fix some code and tests in query assist
> 
> ```diff
> diff --git a/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx b/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx
> index 84ac854c980..e87e74ce299 100644
> --- a/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx
> +++ b/src/plugins/query_enhancements/public/query_assist/components/query_assist_bar.tsx
> @@ -45,7 +45,7 @@ export const QueryAssistBar: React.FC<QueryAssistInputProps> = (props) => {
>      const subscription = props.connectionsService
>        .getSelectedConnection$()
>        .subscribe((connection) => {
> -        dataSourceIdRef.current = connection?.id;
> +        dataSourceIdRef.current = connection?.dataSource.id;
>        });
>      return () => subscription.unsubscribe();
>    }, [props.connectionsService]);
> diff --git a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx
> index ea568959152..41fc36dd71c 100644
> --- a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx
> +++ b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.test.tsx
> @@ -24,13 +24,10 @@ const httpMock = coreSetupMock.http;
>  
>  jest.mock('../components', () => ({
>    QueryAssistBar: jest.fn(() => <div>QueryAssistBar</div>),
> -}));
> -
> -jest.mock('../components/query_assist_banner', () => ({
>    QueryAssistBanner: jest.fn(() => <div>QueryAssistBanner</div>),
>  }));
>  
> -describe.skip('CreateExtension', () => {
> +describe('CreateExtension', () => {
>    afterEach(() => {
>      jest.clearAllMocks();
>    });
> diff --git a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx
> index bd990b57f5a..8d51cb8998c 100644
> --- a/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx
> +++ b/src/plugins/query_enhancements/public/query_assist/utils/create_extension.tsx
> @@ -28,7 +28,7 @@ const getAvailableLanguages$ = (
>    connectionsService.getSelectedConnection$().pipe(
>      distinctUntilChanged(),
>      switchMap(async (connection) => {
> -      const dataSourceId = connection?.id;
> +      const dataSourceId = connection?.dataSource.id;
>        const cached = availableLanguagesByDataSource.get(dataSourceId);
>        if (cached !== undefined) return cached;
>        const languages = await http
> @@ -102,6 +102,7 @@ interface QueryAssistWrapperProps {
>  
>  const QueryAssistWrapper: React.FC<QueryAssistWrapperProps> = (props) => {
>    const [visible, setVisible] = useState(false);
>  
>    useEffect(() => {
>      let mounted = true;
> @@ -112,6 +113,7 @@ const QueryAssistWrapper: React.FC<QueryAssistWrapperProps> = (props) => {
>        props.http
>      ).subscribe((languages) => {
>        const available = languages.includes(props.dependencies.language);
>        if (mounted) setVisible(props.invert ? !available : available);
>      });
>  
> diff --git a/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts b/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts
> index f22bdb14836..5e31191c9c9 100644
> --- a/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts
> +++ b/src/plugins/query_enhancements/server/routes/query_assist/agents.test.ts
> @@ -5,25 +5,29 @@
>  
>  import { ApiResponse } from '@opensearch-project/opensearch';
>  import { ResponseError } from '@opensearch-project/opensearch/lib/errors';
> +import { loggerMock } from '@osd/logging/target/mocks';
>  import { RequestHandlerContext } from 'src/core/server';
> -// // eslint-disable-next-line @osd/eslint/no-restricted-paths
> -// import { CoreRouteHandlerContext } from 'src/core/server/core_route_handler_context';
>  import { coreMock } from '../../../../../core/server/mocks';
> -import { loggerMock } from '@osd/logging/target/mocks';
>  import { getAgentIdByConfig, requestAgentByConfig } from './agents';
>  
> -describe.skip('Agents helper functions', () => {
> -  // const coreContext = new CoreRouteHandlerContext(
> -  //   coreMock.createInternalStart(),
> -  //   httpServerMock.createOpenSearchDashboardsRequest()
> -  // );
> +describe('Agents helper functions', () => {
>    const coreContext = coreMock.createRequestHandlerContext();
>    const client = coreContext.opensearch.client.asCurrentUser;
>    const mockedTransport = client.transport.request as jest.Mock;
>    const context: RequestHandlerContext = {
>      core: coreContext,
> +    // @ts-ignore
>      dataSource: jest.fn(),
> -    query_assist: { dataSourceEnabled: false, logger: loggerMock.create() },
> +    query_assist: {
> +      dataSourceEnabled: false,
> +      logger: loggerMock.create(),
> +      configPromise: Promise.resolve({
> +        enabled: true,
> +        queryAssist: {
> +          supportedLanguages: [{ language: 'PPL', agentConfig: 'testConfig' }],
> +        },
> +      }),
> +    },
>    };
>  
>    afterEach(() => {
> diff --git a/src/plugins/query_enhancements/server/types.ts b/src/plugins/query_enhancements/server/types.ts
> index 2ab716b9892..1a5d962e705 100644
> --- a/src/plugins/query_enhancements/server/types.ts
> +++ b/src/plugins/query_enhancements/server/types.ts
> @@ -51,7 +51,10 @@ export interface FacetRequest {
>    };
>  }
>  
> -declare module '../../../core/server' {
> +// TODO declare it in core changes the type for every reference. we only need
> +// this declaration in query_enhancements, but it doesn't seem possible
> +// https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4274
> +/* declare module '../../../core/server' {
>    interface RequestHandlerContext {
>      query_assist: {
>        logger: Logger;
> @@ -64,4 +67,4 @@ declare module '../../../core/server' {
>        dataSourceEnabled: boolean;
>      };
>    }
> -}
> +} */
> ```

thx joshua, I merged this PR into main did you want to open a pr for that fix into the main repo?

**Comment by joshuali925** [comment]
i added a PR for the skipped tests https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7272

**Pull request merged**